sdcm/cluster.py

```diff
@@ -463,40 +422,190 @@ def wait_cs_installed(self, verbose=True):
+
+    def _wait_public_ip(self):
+        while self._instance.public_ip_address is None:
+            time.sleep(1)
+            self._instance.reload()
+
+    def restart(self):
+        self._instance.stop()
+        self._instance_wait_safe(self._instance.wait_until_stopped)
+        self._instance.start()
+        self._instance_wait_safe(self._instance.wait_until_running)
+        self._wait_public_ip()
+        self.log.debug('Got new public IP %s',
+                       self._instance.public_ip_address)
+        self.remoter.hostname = self._instance.public_ip_address
+        self.wait_db_up()
+
+    def destroy(self):
+        self._instance.terminate()
+        global EC2_INSTANCES
+        EC2_INSTANCES.remove(self._instance)
+        self.log.info('Destroyed')
+
+
+class LibvirtNode(BaseNode):
+
+    """
+    Wraps a domain object, so that we can also control the it through SSH.
+    """
+
+    def __init__(self, domain, hypervisor, node_prefix='node', node_index=1,
+                 domain_username='root', domain_password='', base_logdir=None):
+        name = '%s-%s' % (node_prefix, node_index)
+        self._backing_image = None
+        self._domain = domain
+        self._hypervisor = hypervisor
+        wait.wait_for(self._domain.isActive)
+        self._wait_public_ip()
+        ssh_login_info = {'hostname': self.public_ip_address,
+                          'user': domain_username,
+                          'password': domain_password}
+        super(LibvirtNode, self).__init__(name=name,
+                                          ssh_login_info=ssh_login_info,
+                                          base_logdir=base_logdir)
+
+    def _get_public_ip_address(self):
+        desc = etree.fromstring(self._domain.XMLDesc(0))
+        mac_path = "devices/interface[@type='bridge']/mac"
+        node = desc.find(mac_path)
+        if node is None:
+            return None
+
+        mac = node.get("address")
+        if mac is None:
+            return None
+
+        mac = mac.lower().strip()
+        output = subprocess.Popen(["arp", "-n"],
+                                  stdout=subprocess.PIPE).communicate()[0]
+        lines = [line.split() for line in output.split("\n")[1:]]
+        addresses = [line[0] for line in lines if (line and (line[2] == mac))]
+        if addresses:
+            # Just return the first address, this is a best effort attempt
+            return addresses[0]
+
+    @property
+    def public_ip_address(self):
+        return self._get_public_ip_address()
+
+    @property
+    def private_ip_address(self):
+        return self._get_public_ip_address()
+
+    def _wait_public_ip(self):
+        while self._get_public_ip_address() is None:
+            time.sleep(1)
+
+    # Remove after node setup is finished
+    def db_up(self):
+        try:
+            result = self.remoter.run('netstat -a | grep :9042',
+                                      verbose=False, ignore_status=True)
```

Comment:
How about below strict checking?

    netstat -an|grep ":9042 *\W* *LISTEN"

If the port is configurable by user, can should not hardcode it. I guess it's fixed in our test environment.